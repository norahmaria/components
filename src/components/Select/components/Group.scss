@use '../../../scss/variables' as *;

.select-nm > .options-nm > .group-nm + .group-nm,
.select-nm > .options-nm > .group-nm ~ .group-nm {
  ul {
    border-bottom: none;
    margin-bottom: 0;
    padding-bottom: 0;
  }
}

.select-nm > .options-nm > .group-nm:first-of-type {
  ul {
    padding-top: 1.25rem;
  }
}

.select-nm > .options-nn .option-nm + .group-nm {
  margin-top: 0.75rem;
}

.select-nm > .options-nm > .group-nm {
  ul {
    display: grid;
    gap: 0.35rem;
    width: 100%;

    .title-nm {
      display: flex;
      gap: 0.3rem;
      align-items: center;

      font-size: 0.875em;
      padding: 0 0.5em 0.5em 0.5rem;
      overflow: hidden;

      text-transform: uppercase;
      font-weight: 600;

      color: inherit;
      color: rgba($black, 0.65);

      svg {
        transform: rotate(180deg);

        path,
        rect {
          fill: rgba($black, 0.45);
        }
      }
    }

    .option-nm {
      &:last-of-type {
        padding-bottom: 0.25rem;
      }
    }
  }
}
