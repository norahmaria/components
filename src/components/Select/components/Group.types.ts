import React from 'react'

interface GroupProps {
  title: string
  children: React.ReactChild | React.ReactChild[]
}

export default GroupProps
