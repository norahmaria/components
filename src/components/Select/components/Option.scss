@use '../../../scss/variables' as *;

.select-nm > .options > .option,
.select-nm > .options > .group .option {
  display: flex;
  justify-content: space-between;
  align-items: center;
  list-style: none;
  gap: 0.5rem;

  border: 1px solid transparent;
  border-radius: $br;

  padding: 0.4em 0.5em;

  transition: background 0.25s ease-in-out, color 0.25s ease-in-out,
    border 0.25s ease-in-out, padding 0.25s ease-in-out;
  cursor: pointer;

  .label {
    width: 100%;
    display: flex;
    align-items: center;
    gap: 0.5rem;
  }

  &,
  &[aria-selected='true'] {
    &:hover,
    &:focus-visible,
    &:active {
      outline: none;
      background: darken($white, 5%);
      border: 1px solid lighten($black, 75.5%);
      padding-left: 1em;
    }

    &:focus-visible {
      background: lighten($black, 77.5%);
    }
  }

  &[aria-disabled='true'] {
    background: darken($white, 3%);
    color: #898a8a;
    cursor: not-allowed;

    &:hover {
      padding: 0.4em 0.5em;
    }

    svg {
      opacity: 0.45;
      filter: grayscale(100%);
    }
  }

  &[aria-selected='true'] {
    font-weight: 500;
  }

  &:first-child {
    margin-top: 0.9rem;
  }

  &:last-child {
    margin-bottom: 0.9rem;
  }
}

@each $name, $color in $colors {
  @if $name == 'neutral' {
    .select-nm.#{$name} > .options > .option,
    .select-nm.#{$name} > .options > .group .option {
      &[aria-selected='true'] {
        background: scale-color($color, $saturation: 5%, $lightness: 50%);
        border: 1px solid scale-color($color, $saturation: 5%, $lightness: 50%);
      }
    }
  } @else {
    .select-nm.#{$name} > .options > .option,
    .select-nm.#{$name} > .options > .group .option {
      &[aria-selected='true'] {
        background: scale-color($color, $saturation: 5%, $lightness: 90%);
        border: 1px solid scale-color($color, $saturation: 5%, $lightness: 90%);
      }
    }
  }
}
