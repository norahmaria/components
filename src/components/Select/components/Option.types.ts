interface OptionProps {
  value: string | number
  leftIcon?: JSX.Element | string
  rightIcon?: JSX.Element | string
  disabled?: boolean
  children: string
}

export default OptionProps
