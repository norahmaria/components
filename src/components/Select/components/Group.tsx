import React from 'react'
import './Group.scss'

import { ReactComponent as LineArrow } from '../../../assets/Arrow_Line.svg'
import GroupProps from '../components/Group.types'

const Group: React.FC<GroupProps> = ({ title, ...props }) => {
  return (
    <ul className="group" {...props}>
      <div className="title">
        {/* <LineArrow /> */}
        {title}
      </div>
      {props.children}
    </ul>
  )
}

export default Group
