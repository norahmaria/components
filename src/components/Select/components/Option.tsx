import React from 'react'
import './Option.scss'

import OptionProps from './Option.types'

const Option = ({}: OptionProps) => <></>

export default Option
