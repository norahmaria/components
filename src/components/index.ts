export { default as Button } from './Button/Button'
export { default as Checkbox } from './Checkbox/Checkbox'

import SelectComponent from './Select/Select'
import Option from './Select/components/Option'
import Group from './Select/components/Group'

import SelectProps from './Select/Select.types'
import OptionProps from './Select/components/Option.types'
import GroupProps from './Select/components/Group.types'

type SelectWithChildren = React.FC<SelectProps> & {
  Option: React.FC<OptionProps>
  Group: React.FC<GroupProps>
}

const Select: SelectWithChildren = SelectComponent as SelectWithChildren
Select.Option = Option
Select.Group = Group

export { Select }
