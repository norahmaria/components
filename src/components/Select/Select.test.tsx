import React from 'react'
import { render, act } from '@testing-library/react'

import { Select } from '../../components/index'

type Props = {
  multiple?: boolean
  disabled?: boolean
  disableFirstOption?: boolean
  status?: { type: 'error' | 'warning'; message: string }
  onSelectionChange?: () => void
} | null

const base = ({
  multiple = false,
  disabled = false,
  disableFirstOption = false,
  onSelectionChange = () => {},
  status,
}: Props) => (
  <Select
    status={status}
    multiple={multiple}
    disabled={disabled}
    label="Country"
    onSelectionChange={onSelectionChange}
  >
    <Select.Option disabled={disableFirstOption} value="us">
      United States
    </Select.Option>

    <Select.Group title="Scandinavia">
      <Select.Option value="no">Norway</Select.Option>
      <Select.Option value="se">Sweden</Select.Option>
    </Select.Group>
  </Select>
)

describe('Select Component', () => {
  it('Can render correctly, closed and open', () => {
    const { getByTestId } = render(base({}))

    const select = getByTestId('select')
    expect(select).toMatchSnapshot()

    select.click()
    expect(select).toMatchSnapshot()
  })

  it('Can open select options', () => {
    const { getByRole, getAllByRole } = render(base({}))

    const [toggleButton] = getAllByRole('button')
    const options = getByRole('listbox')

    expect(toggleButton).not.toHaveClass('open-true')
    expect(toggleButton).not.toHaveAttribute('aria-expanded', 'true')

    toggleButton.click()

    expect(toggleButton).toHaveClass('open-true')
    expect(toggleButton).toHaveAttribute('aria-expanded', 'true')

    expect(options).toHaveClass('open-true')
  })

  it('Groups options properly', () => {
    const { getByTestId } = render(base({}))

    expect(getByTestId('select-group').children).toHaveLength(3)

    expect(getByTestId('select-group').children[0]).toHaveTextContent(
      'Scandinavia'
    )
    expect(getByTestId('select-group').children[1]).toHaveTextContent('Norway')
    expect(getByTestId('select-group').children[2]).toHaveTextContent('Sweden')
  })

  it('Can select single', () => {
    const { getAllByRole } = render(base({}))

    const [toggleButton] = getAllByRole('button')
    const [firstOption] = getAllByRole('option')

    toggleButton.click()
    firstOption.click()

    expect(toggleButton).toHaveTextContent('United States')
  })

  it('Can select multiple', () => {
    const { getAllByRole } = render(base({ multiple: true }))

    const [toggleButton] = getAllByRole('button')
    const [firstOption, secondOption] = getAllByRole('option')

    toggleButton.click()
    firstOption.click()
    secondOption.click()

    expect(toggleButton).toHaveTextContent('United States')
    expect(toggleButton).toHaveTextContent('Norway')
  })

  it('Can disable select', () => {
    const onClick = jest.fn()

    const { getAllByRole } = render(base({ disabled: true }))

    const [toggleButton] = getAllByRole('button')
    toggleButton.onclick = onClick

    expect(toggleButton).toBeDisabled()

    toggleButton.click()
    expect(onClick).not.toHaveBeenCalled()
  })

  it('Can disable individual options', () => {
    const { getAllByRole } = render(base({ disableFirstOption: true }))

    const [toggleButton] = getAllByRole('button')
    const [disabledOption] = getAllByRole('option')

    toggleButton.click()

    expect(disabledOption).toHaveAttribute('aria-disabled', 'true')
    expect(disabledOption.tabIndex).toEqual(-1)

    disabledOption.click()
    expect(toggleButton).not.toHaveTextContent('United States')
  })

  it('Can clear selected', () => {
    const { getByTestId, getAllByRole } = render(base({}))

    const [toggleButton] = getAllByRole('button')
    const firstOption = getAllByRole('option')[0]

    toggleButton.click()
    firstOption.click()

    expect(toggleButton).toHaveTextContent('United States')

    const clearIcon = getByTestId('close-icon')
    expect(clearIcon).toBeInTheDocument()

    act(() => {
      clearIcon.dispatchEvent(new MouseEvent('click', { bubbles: true }))
    })

    expect(toggleButton).toHaveTextContent('Country')
    expect(toggleButton).not.toHaveTextContent('United States')
  })

  it('Displays status messages', () => {
    const { getByText, getByTestId } = render(
      base({ status: { type: 'error', message: 'Must select one' } })
    )

    const status = getByText('Must select one')
    const select = getByTestId('select')

    expect(status).toHaveClass('status')
    expect(select).toHaveClass('error')
  })

  it('Runs onSelectionChange when selection changes', () => {
    const onSelectionChange = jest.fn()

    const { getAllByRole, getByTestId } = render(base({ onSelectionChange }))

    const [toggleButton] = getAllByRole('button')
    toggleButton.click()

    const [firstOption, secondOption] = getAllByRole('option')
    firstOption.click()
    secondOption.click()

    const clearIcon = getByTestId('close-icon')

    act(() => {
      clearIcon.dispatchEvent(new MouseEvent('click', { bubbles: true }))
    })

    expect(onSelectionChange).toBeCalledTimes(4)
  })
})
