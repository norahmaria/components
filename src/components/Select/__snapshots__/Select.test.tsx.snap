// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select Component Can render correctly, closed and open 1`] = `
<div
  class="
        select-nm 
        size-medium 
        color-primary 
         
        undefined
      "
  data-testid="select"
>
  <label
    class="form-label-nm disabled-false"
  >
    Country
  </label>
  <button
    aria-disabled="false"
    aria-errormessage="error"
    aria-expanded="false"
    aria-haspopup="listbox"
    aria-invalid="false"
    class="
        select-btn-nm
        input-base-nm
        open-false 
        size-medium 
        color-primary
        includes-tags-false
         
      "
    type="button"
  >
    <div
      class="label-nm"
    >
      <div
        class="placeholder-nm"
      >
        Placeholder
      </div>
    </div>
    <svg
      aria-label="Open / Close"
      class="arrow-icon-nm"
    />
  </button>
  <ul
    aria-label="Country"
    aria-multiselectable="false"
    class="options-nm open-false"
    role="listbox"
    tabindex="-1"
  >
    <li
      aria-disabled="false"
      aria-selected="false"
      class="options-nm__option"
      role="option"
      tabindex="-1"
      value="us"
    >
      <div
        class="options-nm__option-label"
      >
        United States
      </div>
    </li>
    <li
      class="select-nm__group"
    >
      <ul
        aria-label="Scandinavia"
        data-testid="select-group"
        role="listbox"
      >
        <li
          class="select-nm__title"
        >
          Scandinavia
        </li>
        <li
          aria-disabled="false"
          aria-selected="false"
          class="options-nm__option"
          role="option"
          tabindex="-1"
          value="no"
        >
          <div
            class="options-nm__option-label"
          >
            Norway
          </div>
        </li>
        <li
          aria-disabled="false"
          aria-selected="false"
          class="options-nm__option"
          role="option"
          tabindex="-1"
          value="se"
        >
          <div
            class="options-nm__option-label"
          >
            Sweden
          </div>
        </li>
      </ul>
    </li>
  </ul>
</div>
`;

exports[`Select Component Can render correctly, closed and open 2`] = `
<div
  class="
        select-nm 
        size-medium 
        color-primary 
         
        undefined
      "
  data-testid="select"
>
  <label
    class="form-label-nm disabled-false"
  >
    Country
  </label>
  <button
    aria-disabled="false"
    aria-errormessage="error"
    aria-expanded="false"
    aria-haspopup="listbox"
    aria-invalid="false"
    class="
        select-btn-nm
        input-base-nm
        open-false 
        size-medium 
        color-primary
        includes-tags-false
         
      "
    type="button"
  >
    <div
      class="label-nm"
    >
      <div
        class="placeholder-nm"
      >
        Placeholder
      </div>
    </div>
    <svg
      aria-label="Open / Close"
      class="arrow-icon-nm"
    />
  </button>
  <ul
    aria-label="Country"
    aria-multiselectable="false"
    class="options-nm open-false"
    role="listbox"
    tabindex="-1"
  >
    <li
      aria-disabled="false"
      aria-selected="false"
      class="options-nm__option"
      role="option"
      tabindex="-1"
      value="us"
    >
      <div
        class="options-nm__option-label"
      >
        United States
      </div>
    </li>
    <li
      class="select-nm__group"
    >
      <ul
        aria-label="Scandinavia"
        data-testid="select-group"
        role="listbox"
      >
        <li
          class="select-nm__title"
        >
          Scandinavia
        </li>
        <li
          aria-disabled="false"
          aria-selected="false"
          class="options-nm__option"
          role="option"
          tabindex="-1"
          value="no"
        >
          <div
            class="options-nm__option-label"
          >
            Norway
          </div>
        </li>
        <li
          aria-disabled="false"
          aria-selected="false"
          class="options-nm__option"
          role="option"
          tabindex="-1"
          value="se"
        >
          <div
            class="options-nm__option-label"
          >
            Sweden
          </div>
        </li>
      </ul>
    </li>
  </ul>
</div>
`;
